packages/bot/src/utils/firestore.js

```diff
@@ -0,0 +1,152 @@
+            console.error("Error initializing firestore ", error);
+        }
+    }
+
+    const onRunBot = (login_id) => {
+        try {
+            const start_time = server_time.unix();
+            const strategy = Blockly.Xml.domToText(Blockly.Xml.workspaceToDom(Blockly.derivWorkspace));
+            runs.add({
+                start_time,
+                account_id: login_id,
+                xml: strategy,
+            })
+                .then(function (docRef) {
+                    doc_id = docRef.id;
+                })
+                .catch(function (error) {
+                    console.error("Error adding document to firestore ", error);
+                    doc_id = Math.floor(1000 + Math.random() * 9000)
+                });
+        }
+        catch (error) {
+            console.error("Error adding document to firestore when bot runs ", error);
+        }
+    }
+
+    const onStopBot = () => {
+        try {
+            runs.doc(doc_id).update({
+                end_time: server_time.unix(),
+            });
+        }
+        catch (error) {
+            console.error("Error adding document to firestore when bot stops ", error);
+        }
+    }
+
+    const onSummaryChanged = (summary) => {
+        try {
+            if (summary) {
+                runs.doc(doc_id).collection('Summaries').add({
+                    lost_contracts: summary.lost_contracts,
+                    number_of_runs: summary.number_of_runs,
+                    total_profit: summary.total_profit,
+                    total_payout: summary.total_payout,
+                    total_stake: summary.total_stake,
+                    won_contracts: summary.won_contracts,
+                    time_stamp: server_time.unix()
+                });
+            }
+        }
+        catch (error) {
+            console.error("Error adding document to firestore when summary changes ", error);
+        }
+    }
+
+
+    const onTransactionClosed = (contracts) => {
+        try {
+            const contract = contracts.length > 0 && contracts[0];
+            if (contract.is_completed) {
+                runs.doc(doc_id).collection('Transactions').add({
+                    buy_price: contract.buy_price,
+                    contract_type: contract.contract_type,
+                    currency: contract.currency,
+                    refrence_id: contract.refrence_id,
+                    entry_spot: contract.entry_spot,
+                    exit_spot: contract.exit_spot,
+                    profit: contract.profit,
+                    time_stamp: server_time.unix()
+                });
+            }
+        }
+        catch (error) {
+            console.error("Error adding document to firestore when transaction closes", error);
+        }
+    }
+
+
+    const onErrorHappend = (messages) => {
+        try {
+            const item = messages.length > 0 && messages[0];
+            if (item.message_type === message_types.ERROR) {
```

Comment:
Same like with `contract` earlier in this file, if `item` is `false` you'll get a `TypeError` when trying to access the property. You can fix that by doing something like `if (item && item.message_type === message_types.ERROR)`